_posts/2020-07-07-aws-localstack.md

```diff
@@ -0,0 +1,252 @@
+
+The AWS SDK (Software Development Kit) and CLI are an integral part of our toolset for building applications with AWS services. The SDK provides client libraries in all the popular programming languages like java, node js, python for accessing various AWS services. It essentially does the heavy-lifting of securely connecting to the AWS services.
+
+Both the AWS SDK and the CLI give an option of overriding the URL of AWS API in the local environment for connecting to LocalStack.
+We do this in AWS CLI using commands like this:
+```
+aws --endpointurl http://localhost:4956 kinesis list-streams
+```
+I am connecting to LocalStack on port 4956.
+
+Similarly in SDK our code looks will resemble the below snippet - 
+```java
+URI endpointOverride = new URI("http://localhost:4566");
+S3Client s3 = S3Client.builder().endpointOverride(endpointOverride ).region(region).build();
+```
+
+## How To Run LocalStack
+
+We run LocalStack either as a python application using our local python installation or alternately using a docker image. 
+
+### Run with python
+```
+pip install localstack
+```
+```
+localstack start
+```
+This starts a docker container with the below output.
+
+[![Console](/assets/img/posts/aws-localstack/localstackconsole.png)](/assets/img/posts/aws-localstack/localstackconsole.png)
+
+ We can also run as docker image wuth the docker run command or docker-compose.
+
+
+## Common Usages
+
+#### Create Fake Profile
+We start in the usual way by creating a profile with any arbitrary credentials.
+```
+aws configure --profile localstack
+```
+Here we create a profile named localstack. You can use any other name. 
+Provide any value you like for aws Access Key and Secret Access Key and a valid AWS region like us-east-1 except leaving them blank. Unlike AWS, LocalStack does not validate these credentials but complains if no profile is set. So far, it is just any other AWS profile which we will use to work with LocalStack.
+
+#### Execute CLI Commands
+With our profile created, we execute CLI commands by passing an additional parameter for overriding the endpoint URL as shown here.
+```
+aws s3 --endpoint-url http://localhost:4566 create-bucket io.pratik.mybucket
+```
+***We create an S3 bucket in LocalStack by overriding the endpoint URL.***
+Next, we execute a regular cloudformation template to create a stack composed of an S3 bucket and a DynamoDB table.
+
+```
+aws cloudformation --endpoint-url http://localhost:4566 create-stack --stack-name samplestack --template-body file://sample.yaml --profile localstack
+```
+
+```
+aws cloudformation --endpoint-url http://localhost:4566 delete-stack --stack-name samplestack --profile localstack
+```
+
+### Execute JUnit Tests
+While running as JUnit, ***LocalStack is started as a docker container at a random port and stopped after all tests have finished execution.*** 
+
+The code snippet is a JUnit Jupiter test used to test a java class to store an object in an S3 bucket.
```

Comment:
I moved this sentence to after the code snippet. Please explain the annotations you're using in the test.